src/main/java/org/pkcs11/jacknji11/Metrics.java

```diff
@@ -0,0 +1,269 @@
+package org.pkcs11.jacknji11;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static java.util.Collections.synchronizedMap;
+
+public class Metrics {
```

Comment:
Would be great with some class javadoc here

Reply:
of course.

Reply:
Assuming there is a performance hit by using Metrics, and you have to enable/disable it?

Very neat feature btw.

Reply:
In my world I/O latency triumphs over any microseconds this proxying mechanism might add. Also if you know Spring or Hibernate they basically founded on this. For past 20 years I wasn't complaining about it. Also reflection got a lot faster and is pretty optimized to the point of perfection by now.

Reply:
I've only used debug logging on org.pkcs11.jacknji11.Cryptoki to debug log PKCS#11 calls.
I like the idea of the Metrics, it looks very useful and I think I could make use if it.
Imho it would be better to add Metrics as a separate PR as it is another feature than the attribute fetching. I.e. one feature per PR. What do you think @joelhockey ?

Reply:
It would be easier to have each of these commits as a separate PR, but it probably doesn't matter too much now

Reply:
just 2c: if we split metrics in a separate PR then it must go in before the attribute fetching as I'm checking metrics to verify the optimization in test. I think keeping things separate in logical commits is sufficient and easier to manger and I obviously will squash all the changes from the remarks into target commits just before merging.